mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala

```diff
@@ -0,0 +1,300 @@
+        Vectors.dense(getColStatistics(doubleRDD))
+      case _: VectorUDT =>
+        val filteredDF = dataset.where(s"${$(inputCol)} IS NOT NULL").select($(inputCol))
+        val vectorRDD = filteredDF.rdd.map(_.getAs[Vector](0)).cache()
+        val vl = dataset.first().getAs[Vector]($(inputCol)).size
+        $(strategy) match {
+          case "mean" =>
+            val summary = vectorRDD.treeAggregate((new Array[Double](vl), new Array[Int](vl)))(
+              (prev, data) => (prev, data) match { case ((mean, count), data) =>
+                  var i = 0
+                  while (i < mean.length) {
+                    if (data(i) != 0 && !data(i).isNaN) {
+                      count(i) += 1
+                      mean(i) = mean(i) + (data(i) - mean(i)) / count(i)
+                    }
+                    i += 1
+                  }
+                  (mean, count)
+              }, (aggregator1, aggregator2) => (aggregator1, aggregator2) match {
+                case ((mean1, c1), (mean2, c2)) =>
+                  (0 until mean1.length).foreach{ i =>
+                    mean1(i) = mean1(i) + (mean2(i) - mean1(i)) * c2(i) / (c1(i) + c2(i))
+                    c1(i) += c2(i)
+                  }
+                  (mean1, c1)
+              })
+            Vectors.dense(summary._1)
+          case _ =>
+            val statisticsArray = new Array[Double](vl)
+            (0 until vl).foreach(i => {
+              statisticsArray(i) = getColStatistics(vectorRDD.map(v => v(i)))
+            })
+            Vectors.dense(statisticsArray)
+        }
+    }
+    copyValues(new ImputerModel(uid, alternate).setParent(this))
+  }
+
+  /** Extract the statistics info from a Double column according to the strategy */
+  private def getColStatistics(data: RDD[Double]): Double = {
+    val filteredRDD = data.filter(!isMissingValue(_))
+    val colStatistics = $(strategy) match {
+      case "mean" => filteredRDD.mean()
+      case "median" => filteredRDD.sortBy(d => d).zipWithIndex()
+        .map(p => (p._2, p._1)).lookup(filteredRDD.count() / 2).head
+      case "most" => filteredRDD.map(d => (d, 1)).reduceByKey(_ + _).sortBy(-_._2).first()._1
+      case _ => throw new SparkException(s"unsupported impute strategy: ${$(strategy)}")
+    }
+    colStatistics
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema)
+  }
+
+  override def copy(extra: ParamMap): Imputer = {
+    val copied = new Imputer(uid)
+    copyValues(copied, extra)
+  }
+}
+
+@Since("1.6.0")
+object Imputer extends DefaultParamsReadable[Imputer] {
+
+  /** Set of strategy names that Imputer currently supports. */
+  private[ml] val supportedStrategyNames = Set("mean", "median", "most")
+
+  @Since("1.6.0")
+  override def load(path: String): Imputer = super.load(path)
+}
+
+/**
+  * :: Experimental ::
+  * Model fitted by [[Imputer]].
+  *
+  * @param alternate statistics value for each feature during fitting
```

Comment:
Is "alternate" the commonly used name for imputation?  I have heard "surrogate" used more frequently, but may be wrong.